### PR TITLE
display inlay hints with rust-analyzer

### PR DESCRIPTION
**Pull request by brotzeit**
fixes brotzeit/rustic#73

### CODE REVIEW

**Comment by flodiebold** [comment]
Note that the inlay hints interact badly with the lsp-ui sideline, because it doesn't seem to consider the overlays in its width calculation, which often leads to lines wrapping around :confused:

**Comment by brotzeit** [comment]
Oh. Thanks for the info. I will add a note.

File: lsp-rust.el [hunk 1]
```diff
@@ -425,6 +430,57 @@ PARAMS progress report notification data."
...
       (message (format "Switched to server %s." server)))))
 
+;; inlay hints
+
+(defvar-local lsp-rust-analyzer-inlay-hints-timer nil)
+
+(defun lsp-rust-analyzer-update-inlay-hints (buffer)
+  (if (and (lsp-rust-analyzer-initialized?)
+           (eq buffer (current-buffer)))
+      (lsp-send-request-async
```
Comment (yyoncho) [inline lsp-rust.el]: can you switch to lsp-request-async? lsp-send-request-async is deprecated(although there is no indication for that). This is an artefact from old lsp-mode. I will add deprecation notice.

File: lsp-rust.el [hunk 2]
```diff
@@ -425,6 +430,57 @@ PARAMS progress report notification data."
                          (* (lsp--client-priority (gethash server lsp-clients)) -1)))
       (message (format "Switched to server %s." server)))))
 
+;; inlay hints
+
+(defvar-local lsp-rust-analyzer-inlay-hints-timer nil)
+
+(defun lsp-rust-analyzer-update-inlay-hints (buffer)
+  (if (and (lsp-rust-analyzer-initialized?)
```
Comment (yyoncho) [inline lsp-rust.el]: here we should update only if the current buffer has changed(I assume that the hints won't change if there is no change in the buffer content). I filed a task to unify all tasks that run on timer - https://github.com/emacs-lsp/lsp-mode/issues/1148 so you could leave it as it is and once we solve #1148 we may change the code to use the newly introduced hook.

**Comment by yyoncho** [comment]
I would suggest adding also a functions to show/hide the hints one time.

**Comment by marcbowes** [comment]
> 
> 
> Note that the inlay hints interact badly with the lsp-ui sideline, because it doesn't seem to consider the overlays in its width calculation, which often leads to lines wrapping around 😕

Is there a issue tracking this? I was not able to find one. Seems like this is one of the biggest pieces of functionality missing in the out-of-the-box experience (for example, comparing what VSCode gives vs. Emacs).

**Pull request merged**